lib/optimize/LimitChunkCountPlugin.js

```diff
@@ -46,53 +69,174 @@ class LimitChunkCountPlugin {
-							// extend combination pairs with size and integrated size
-							const a = chunkGraph.getChunkSize(pair[0], options);
-							const b = chunkGraph.getChunkSize(pair[1], options);
-							const ab = chunkGraph.getIntegratedChunksSize(
-								pair[0],
-								pair[1],
+					// create a lazy sorted data structure to keep all combinations
+					// this is large. Size = chunks * (chunks - 1) / 2
+					// It uses a multi layer bucket sort plus normal sort in the last layer
+					// It's also lazy so only accessed buckets are sorted
+					const combinations = new LazyBucketSortedSet(
+						// Layer 1: ordered by largest size benefit
+						c => c.sizeDiff,
+						(a, b) => b - a,
+						// Layer 2: ordered by smallest combined size
+						c => c.integratedSize,
+						(a, b) => a - b,
+						// Layer 3: ordered by position difference in orderedChunk (-> to be deterministic)
+						c => c.bIdx - c.aIdx,
+						(a, b) => a - b,
+						// Layer 4: ordered by position in orderedChunk (-> to be deterministic)
+						(a, b) => a.bIdx - b.bIdx
+					);
+
+					// we keep a mappng from chunk to all combinations
+					// but this mapping is not kept up-to-date with deletions
+					// so `deleted` flag need to be considered when iterating this
+					/** @type {Map<Chunk, Set<ChunkCombination>>} */
+					const combinationsByChunk = new Map();
+
+					orderedChunks.forEach((b, bIdx) => {
+						// create combination pairs with size and integrated size
+						for (let aIdx = 0; aIdx < bIdx; aIdx++) {
+							const a = orderedChunks[aIdx];
+							const integratedSize = chunkGraph.getIntegratedChunksSize(
+								a,
+								b,
 								options
 							);
-							/** @type {[number, number, Chunk, Chunk, number, number]} */
-							const extendedPair = [a + b - ab, ab, pair[0], pair[1], a, b];
-							return extendedPair;
-						})
-						.sort((a, b) => {
-							// sadly javascript does an inplace sort here
-							// sort them by size
-							const diff1 = b[0] - a[0];
-							if (diff1 !== 0) return diff1;
-							const diff2 = a[1] - b[1];
-							if (diff2 !== 0) return diff2;
-							const diff3 = compareChunksWithGraph(a[2], b[2]);
-							if (diff3 !== 0) return diff3;
-							return compareChunksWithGraph(a[3], b[3]);
-						});
-
-					const pair = sortedExtendedPairCombinations[0];
-
-					if (pair) {
-						chunkGraph.integrateChunks(pair[2], pair[3]);
-						compilation.chunks.delete(pair[3]);
-						return true;
+
+							// filter pairs that do not have an integratedSize
+							// meaning they can NOT be integrated!
+							if (integratedSize === false) continue;
+
+							const aSize = chunkGraph.getChunkSize(a, options);
+							const bSize = chunkGraph.getChunkSize(b, options);
+							const c = {
+								deleted: false,
+								sizeDiff: aSize + bSize - integratedSize,
+								integratedSize,
+								a,
+								b,
+								aIdx,
+								bIdx,
+								aSize,
+								bSize
+							};
+							combinations.add(c);
```

Comment:
@sokra I am just curious why you use here lazy sorting? Looks like inserting in right order will be faster.

Reply:
Ah, you deleting unprocessed combinations later